datafusion/ffi/src/udaf/accumulator.rs

```diff
@@ -217,9 +225,20 @@ pub struct ForeignAccumulator {
 unsafe impl Send for ForeignAccumulator {}
 unsafe impl Sync for ForeignAccumulator {}
 
-impl From<FFI_Accumulator> for ForeignAccumulator {
-    fn from(accumulator: FFI_Accumulator) -> Self {
-        Self { accumulator }
+impl From<FFI_Accumulator> for Box<dyn Accumulator> {
+    fn from(mut accumulator: FFI_Accumulator) -> Self {
+        if (accumulator.library_marker_id)() == crate::get_library_marker_id() {
+            unsafe {
+                let private_data = Box::from_raw(
+                    accumulator.private_data as *mut AccumulatorPrivateData,
+                );
+                // We must set this to null to avoid a double free
+                accumulator.private_data = null_mut();
```

Comment:
just thinking aloud, would be it still protected from double free in if function called concurrently?